core/trino-main/src/main/java/io/trino/sql/SensitiveStatementRedactor.java

```diff
@@ -0,0 +1,131 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql;
+
+import com.google.inject.Inject;
+import io.trino.FeaturesConfig;
+import io.trino.connector.CatalogFactory;
+import io.trino.spi.connector.ConnectorName;
+import io.trino.sql.tree.AstVisitor;
+import io.trino.sql.tree.CreateCatalog;
+import io.trino.sql.tree.Explain;
+import io.trino.sql.tree.ExplainAnalyze;
+import io.trino.sql.tree.Identifier;
+import io.trino.sql.tree.Node;
+import io.trino.sql.tree.Property;
+import io.trino.sql.tree.Statement;
+import io.trino.sql.tree.StringLiteral;
+
+import java.util.List;
+import java.util.Set;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+
+public class SensitiveStatementRedactor
+{
+    public static final String REDACTED_VALUE = "***";
+
+    private final boolean enabled;
+    private final CatalogFactory catalogFactory;
+
+    @Inject
+    public SensitiveStatementRedactor(FeaturesConfig config, CatalogFactory catalogFactory)
+    {
+        this.enabled = config.isStatementRedactingEnabled();
+        this.catalogFactory = catalogFactory;
+    }
+
+    public String redact(String rawQuery, Statement statement)
+    {
+        if (enabled) {
+            RedactingVisitor visitor = new RedactingVisitor();
+            Node redactedStatement = visitor.process(statement);
+            if (visitor.isRedacted()) {
+                return SqlFormatter.formatSql(redactedStatement);
+            }
+        }
+        return rawQuery;
+    }
+
+    public String redact(String rawQuery)
+    {
+        if (enabled) {
+            return REDACTED_VALUE;
+        }
+        return rawQuery;
+    }
+
+    private class RedactingVisitor
+            extends AstVisitor<Node, Void>
+    {
+        private boolean redacted;
+
+        public boolean isRedacted()
+        {
+            return redacted;
+        }
+
+        @Override
+        protected Node visitNode(Node node, Void context)
+        {
+            return node;
+        }
+
+        @Override
+        protected Node visitExplain(Explain explain, Void context)
+        {
+            Statement statement = (Statement) process(explain.getStatement());
+            return new Explain(explain.getLocation().orElseThrow(), statement, explain.getOptions());
+        }
+
+        @Override
+        protected Node visitExplainAnalyze(ExplainAnalyze explainAnalyze, Void context)
+        {
+            Statement statement = (Statement) process(explainAnalyze.getStatement());
+            return new ExplainAnalyze(explainAnalyze.getLocation().orElseThrow(), statement, explainAnalyze.isVerbose());
+        }
+
+        @Override
+        protected Node visitCreateCatalog(CreateCatalog createCatalog, Void context)
```

Comment:
Is there some way to notice when we need to add new node visitors here?

Should this be a "wrapper" like the various `Forwarding***` classes and a test to assert that full set of methods is overridden? That way once new methods get added we'll explicitly need to either override to do no-op or to redact?

WDYT? Might be overkill for now so need to change anything - just to have a discussion.